ThatDefaultProject/Assets/ThatDefaultProject/Scripts/MonoSingleton.cs

```diff
@@ -0,0 +1,62 @@
+using UnityEngine;
+
+namespace that
+{
+    public class MonoSingleton<T> : MonoBehaviour where T : MonoSingleton<T>
+    {
+        private static bool _isExiting = false;
+        private static T _instance;
+
+        // Will return null if the application is exiting. When requesting Instance in OnDestroy, always perform null checks.
+        public static T Instance
+        {
+            get
+            {
+                if (_isExiting)
+                {
+                    return null;
+                }
+
+                if (_instance == null)
+                {
+                    _instance = FindObjectOfType<T>();
+                    if (_instance == null)
+                    {
+                        GameObject obj = new GameObject();
+                        obj.name = "[S] " + typeof(T).ToString();
+                        _instance = obj.AddComponent<T>();
+                    }
+                }
+                return _instance;
+            }
+        }
+
+        // When override awake, make sure to run base.Awake(), or inherit from LateAwake().
+        protected virtual void Awake()
+        {
+            if (_instance == null)
+            {
+                _instance = this as T;
+                if (transform.parent == null)
+                {
+                    DontDestroyOnLoad(gameObject);
```

Comment:
Do we want every singleton to be available through the whole game?
Won't we sometimes need a singleton only in a game-scene for example.

Reply:
You expect a singleton to be available when it is needed, at least that is what I learned. If you want a singleton not available the entire time you either attach it to a parent that will be deleted (as you can see in the code snippet) or you should look at a different option (service locator perhaps)

If you disagree, I am fine with making a variation that only stays in one scene.

Reply:
My bad, I didn't see that the singleton doesn't get the DontDestroyOnLoad call if it is parented to an object.
This approach is perfect for scene-specific singletons, imo.

Approving this pull-request.